File: Programmers/[17683] 방금그곡/JFe/17683.py
import re
def solution(m, musicinfos):
    answer, length = "(None)", 0
    # '#' 들어간 음 변환
    for org, rep in zip(['A#', 'C#', 'D#', 'F#', 'G#'], ['H', 'I', 'J', 'K', 'L']):
        m = m.replace(org, rep)
    pattern = re.compile(m)
    for info in musicinfos:
        info = info.split(',')
        info[0] = info[0].split(':')
        info[1] = info[1].split(':')
        time = (int(info[1][0])*60 + int(info[1][1])) - (int(info[0][0])*60 + int(info[0][1]))
        # '#' 들어간 음 변환
        for org, rep in zip(['A#', 'C#', 'D#', 'F#', 'G#'], ['H', 'I', 'J', 'K', 'L']):
            info[3] = info[3].replace(org, rep)
        message = ''
        # time만큼 악보 정보 반복해서 message 저장
        for i in range(time):
            message += info[3][i%len(info[3])]
        # pattern이 있고, 현재 저장된 길이보다 크면 answer 변경
        if pattern.search(message) and length < time:
            answer = info[2]
            length = time
    return answer

File: Programmers/[17683] 방금그곡/SUbbb/17683.java
import java.util.*;

class JustNowMusic {
    // Map의 순서 보장을 위해 LinkedHashMap 사용
    LinkedHashMap<String, String> music;

    public JustNowMusic(String[] musicinfos) {
        music = new LinkedHashMap<>();

        for (String mi : musicinfos) initMusic(mi);
    }

    private void initMusic(String mi) {
        String[] info = mi.split(",");

        int runTime = parsing(info[0], info[1]);

        StringBuilder score = new StringBuilder();

        info[3] = convert(info[3]);
        
        // 재생 시간만큼 악보 반복
        for (int i = 0, idx = 0; i < runTime; i++) {
            if (i % info[3].length() == 0) idx = 0;
            score.append(info[3].charAt(idx));
            idx++;
        }

        // 제목과 악보를 Map에 저장
        music.put(info[2], score.toString());
    }
    
    // #이 들어가는 음 치환
    private String convert(String s) {
        return s.replace("C#", "c").replace("D#", "d").replace("F#", "f").replace("G#", "g").replace("A#", "a");
    }

    private int parsing(String info1, String info2) {
        int startTime = getTime(info1);
        int endTime = getTime(info2);

        return endTime - startTime;
    }

    private int getTime(String info) {
        String[] time = info.split(":");
        int hour = Integer.parseInt(time[0]);
        int min = Integer.parseInt(time[1]);
        
        return hour * 60 + min;
    }

    public String findMusic(String m) {
        m = convert(m);
        int maxPlayTime = -1;
        
        String answer = "";
        for (Map.Entry<String, String> entry : music.entrySet()) {
            String score = entry.getValue();
            int playTime = score.length();
            
            // 기억한 멜로디를 포함하고, 재생 시간이 제일 긴 음악인 경우 answer에 해당 제목 저장
            if (score.contains(m) && playTime > maxPlayTime) {
                answer = entry.getKey();
                maxPlayTime = playTime;
            }
        }

        return maxPlayTime == -1? "(None)" : answer;
    }
}

class Solution {
    public String solution(String m, String[] musicinfos) {
        JustNowMusic jnm = new JustNowMusic(musicinfos);
        
        return jnm.findMusic(m);
    }
}